bus-mapping/src/evm/opcodes.rs

```diff
@@ -412,30 +413,42 @@ pub fn gen_begin_tx_ops(state: &mut CircuitInputStateRef) -> Result<ExecStep, Er
     // Get code_hash of callee
     let (_, callee_account) = state.sdb.get_account(&call.address);
     let callee_exists = !callee_account.is_empty();
-    let code_hash = callee_account.code_hash;
-    let callee_code_hash = call.code_hash;
-    let (callee_code_hash_word, is_empty_code_hash) = if callee_exists {
-        debug_assert_eq!(
-            callee_code_hash, code_hash,
-            "call.address {:?} callee_account {:?}",
-            call.address, callee_account
-        );
-
-        (
-            callee_code_hash.to_word(),
-            callee_code_hash.to_fixed_bytes() == *EMPTY_HASH,
-        )
-    } else {
-        (Word::zero(), true)
+    let (callee_code_hash, is_empty_code_hash) = match (state.tx.is_create(), callee_exists) {
+        (true, _) => (call.code_hash.to_word(), false),
+        (_, true) => {
+            debug_assert_eq!(
+                callee_account.code_hash, call.code_hash,
+                "callee account's code hash: {:?}, call's code hash: {:?}",
+                callee_account.code_hash, call.code_hash
+            );
+            (
+                call.code_hash.to_word(),
+                call.code_hash.to_fixed_bytes() == *EMPTY_HASH,
+            )
+        }
+        (_, false) => (Word::zero(), true),
     };
 
-    state.account_read(
-        &mut exec_step,
-        call.address,
-        AccountField::CodeHash,
-        callee_code_hash_word,
-        callee_code_hash_word,
-    )?;
+    if state.tx.is_create() {
+        state.push_op_reversible(
+            &mut exec_step,
+            RW::WRITE,
+            AccountOp {
+                address: call.address,
+                field: AccountField::CodeHash,
+                value: callee_code_hash,
+                value_prev: Word::zero(),
```

Comment:
in theory,  a user can pre-calculate his future contract address, and then transfer some ethers into that address. So when he is doing the deployment, the "call.address" is not non-exist but with some eth balance... For that senario, the line 440 will be empty_code_hash instead of zero. (but i guess this is really edge case, we can either fix that or just add some TODO comments)

Reply:
Actually if you check here: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder/input_state_ref.rs#L281-L300, i.e.
```rust
        let account_value_prev = match op.field {
            ...
            AccountField::CodeHash => {
                if account.is_empty() {
                    Word::zero()
                } else {
                    account.code_hash.to_word()
                }
            }
        };
```

This means, the `account_value_prev` that is later compared against the operation's previous value is set to `0` if the code hash matches EMPTY_HASH. So for both existent/non-existent accounts, as long as the code is not set, the previous value is zero.